# Problem -- 1)

Classify the Size_Categorie using SVM for Forestfires data.

#### Data Description:

month month of the year: 'jan' to 'dec'

day day of the week: 'mon' to 'sun'

FFMC: FFMC index from the FWI system: 18.7 to 96.20

DMC: DMC index from the FWI system: 1.1 to 291.3

DC: DC index from the FWI system: 7.9 to 860.6

ISI: ISI index from the FWI system: 0.0 to 56.10

temp: temperature in Celsius degrees: 2.2 to 33.30

RH: relative humidity in %: 15.0 to 100

wind: wind speed in km/h: 0.40 to 9.40

rain: outside rain in mm/m2 : 0.0 to 6.4

Size_Categorie: the burned area of the forest ( Small , Large)

In [2]:
#Importing Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings('ignore')

from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.metrics import classification_report

from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split,cross_val_score

In [5]:
forest= pd.read_csv('C:/Users/ancha/Downloads/forestfires.csv')
forest

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  ...  monthfeb  \
0     mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0  ...         0   
1     oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0  ...         0   
2     oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0  ...         0   
3     mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2  ...         0   
4     mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0  ...         0   
..    ...  ...   ...    ...    ...   ...   ...  ..   ...   ...  ...       ...   
512   aug  sun  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0  ...         0   
513   aug  sun  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  ...         0   
514   aug  sun  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  ...         0   
515   aug  sat  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0  ...         0   
516   nov  tue  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0  ...         0   

     monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0           0         0         0         1         0         0         0   
1           0         0         0         0         0         0         1   
2           0         0         0         0         0         0         1   
3           0         0         0         1         0         0         0   
4           0         0         0         1         0         0         0   
..        ...       ...       ...       ...       ...       ...       ...   
512         0         0         0         0         0         0         0   
513         0         0         0         0         0         0         0   
514         0         0         0         0         0         0         0   
515         0         0         0         0         0         0         0   
516         0         0         0         0         0         1         0   

     monthsep  size_category  
0           0          small  
1           0          small  
2           0          small  
3           0          small  
4           0          small  
..        ...            ...  
512         0          large  
513         0          large  
514         0          large  
515         0          small  
516         0          small  

[517 rows x 31 columns]

# EDA

In [6]:
forest.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 31 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   month          517 non-null    object 
 1   day            517 non-null    object 
 2   FFMC           517 non-null    float64
 3   DMC            517 non-null    float64
 4   DC             517 non-null    float64
 5   ISI            517 non-null    float64
 6   temp           517 non-null    float64
 7   RH             517 non-null    int64  
 8   wind           517 non-null    float64
 9   rain           517 non-null    float64
 10  area           517 non-null    float64
 11  dayfri         517 non-null    int64  
 12  daymon         517 non-null    int64  
 13  daysat         517 non-null    int64  
 14  daysun         517 non-null    int64  
 15  daythu         517 non-null    int64  
 16  daytue         517 non-null    int64  
 17  daywed         517 non-null    int64  
 18  monthapr  

In [7]:
forest.describe()

FFMC         DMC          DC         ISI        temp          RH  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean    90.644681  110.872340  547.940039    9.021663   18.889168   44.288201   
std      5.520111   64.046482  248.066192    4.559477    5.806625   16.317469   
min     18.700000    1.100000    7.900000    0.000000    2.200000   15.000000   
25%     90.200000   68.600000  437.700000    6.500000   15.500000   33.000000   
50%     91.600000  108.300000  664.200000    8.400000   19.300000   42.000000   
75%     92.900000  142.400000  713.900000   10.800000   22.800000   53.000000   
max     96.200000  291.300000  860.600000   56.100000   33.300000  100.000000   

             wind        rain         area      dayfri  ...    monthdec  \
count  517.000000  517.000000   517.000000  517.000000  ...  517.000000   
mean     4.017602    0.021663    12.847292    0.164410  ...    0.017408   
std      1.791653    0.295959    63.655818    0.371006  ...    0.130913   
min      0.400000    0.000000     0.000000    0.000000  ...    0.000000   
25%      2.700000    0.000000     0.000000    0.000000  ...    0.000000   
50%      4.000000    0.000000     0.520000    0.000000  ...    0.000000   
75%      4.900000    0.000000     6.570000    0.000000  ...    0.000000   
max      9.400000    6.400000  1090.840000    1.000000  ...    1.000000   

         monthfeb    monthjan    monthjul    monthjun    monthmar    monthmay  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean     0.038685    0.003868    0.061896    0.032882    0.104449    0.003868   
std      0.193029    0.062137    0.241199    0.178500    0.306138    0.062137   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

         monthnov    monthoct    monthsep  
count  517.000000  517.000000  517.000000  
mean     0.001934    0.029014    0.332689  
std      0.043980    0.168007    0.471632  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000  
75%      0.000000    0.000000    1.000000  
max      1.000000    1.000000    1.000000  

[8 rows x 28 columns]

In [8]:
forest.duplicated().value_counts()

False    509
True       8
dtype: int64

In [9]:
#Dropping the duplicated values.
forest2= forest.copy()
forest2= forest.drop_duplicates()
forest2

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  ...  monthfeb  \
0     mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0  ...         0   
1     oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0  ...         0   
2     oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0  ...         0   
3     mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2  ...         0   
4     mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0  ...         0   
..    ...  ...   ...    ...    ...   ...   ...  ..   ...   ...  ...       ...   
512   aug  sun  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0  ...         0   
513   aug  sun  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  ...         0   
514   aug  sun  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  ...         0   
515   aug  sat  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0  ...         0   
516   nov  tue  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0  ...         0   

     monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0           0         0         0         1         0         0         0   
1           0         0         0         0         0         0         1   
2           0         0         0         0         0         0         1   
3           0         0         0         1         0         0         0   
4           0         0         0         1         0         0         0   
..        ...       ...       ...       ...       ...       ...       ...   
512         0         0         0         0         0         0         0   
513         0         0         0         0         0         0         0   
514         0         0         0         0         0         0         0   
515         0         0         0         0         0         0         0   
516         0         0         0         0         0         1         0   

     monthsep  size_category  
0           0          small  
1           0          small  
2           0          small  
3           0          small  
4           0          small  
..        ...            ...  
512         0          large  
513         0          large  
514         0          large  
515         0          small  
516         0          small  

[509 rows x 31 columns]

In [10]:
#Frequency count.
for var in forest2:
    print(forest2[var].value_counts())

aug    178
sep    172
mar     53
jul     32
feb     20
jun     16
oct     15
apr      9
dec      9
jan      2
may      2
nov      1
Name: month, dtype: int64
sun    94
fri    83
sat    82
mon    74
tue    63
thu    60
wed    53
Name: day, dtype: int64
92.1    27
91.6    27
91.0    21
91.7    18
92.4    16
        ..
86.9     1
85.2     1
53.4     1
85.1     1
94.4     1
Name: FFMC, Length: 106, dtype: int64
99.0     10
129.5     9
231.1     7
137.0     7
108.3     7
         ..
138.1     1
109.2     1
91.6      1
122.3     1
146.0     1
Name: DMC, Length: 215, dtype: int64
745.3    10
692.6     9
698.6     8
692.3     8
706.4     7
         ..
565.5     1
503.6     1
589.9     1
85.3      1
106.7     1
Name: DC, Length: 219, dtype: int64
9.6     22
7.1     19
6.3     19
7.0     17
8.4     16
        ..
16.5     1
2.2      1
8.3      1
2.0      1
1.1      1
Name: ISI, Length: 119, dtype: int64
17.4    8
19.6    8
20.6    7
15.4    7
21.9    6
       ..
16.3    1
12.8    1
10.4    1
8.7 

# Label Encoding

In [11]:
from sklearn import preprocessing

forest2_lab= forest2.copy()

for i, column in enumerate(forest2_lab[[feature for feature in forest2_lab.columns
        if forest2_lab[feature].dtypes == 'O']].columns, 1):
    
    
    label_encoder= preprocessing.LabelEncoder()
    forest2_lab[column]=label_encoder.fit_transform(forest2_lab[column])
forest2_lab
        

month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  ...  \
0        7    0  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0  ...   
1       10    5  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0  ...   
2       10    2  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0  ...   
3        7    0  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2  ...   
4        7    3  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0  ...   
..     ...  ...   ...    ...    ...   ...   ...  ..   ...   ...  ...   
512      1    3  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0  ...   
513      1    3  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  ...   
514      1    3  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  ...   
515      1    2  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0  ...   
516      9    5  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0  ...   

     monthfeb  monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  \
0           0         0         0         0         1         0         0   
1           0         0         0         0         0         0         0   
2           0         0         0         0         0         0         0   
3           0         0         0         0         1         0         0   
4           0         0         0         0         1         0         0   
..        ...       ...       ...       ...       ...       ...       ...   
512         0         0         0         0         0         0         0   
513         0         0         0         0         0         0         0   
514         0         0         0         0         0         0         0   
515         0         0         0         0         0         0         0   
516         0         0         0         0         0         0         1   

     monthoct  monthsep  size_category  
0           0         0              1  
1           1         0              1  
2           1         0              1  
3           0         0              1  
4           0         0              1  
..        ...       ...            ...  
512         0         0              0  
513         0         0              0  
514         0         0              0  
515         0         0              1  
516         0         0              1  

[509 rows x 31 columns]

# Standardizing and Data Splitting

In [12]:
ss= StandardScaler()
forest_ss= forest2_lab.copy()

In [13]:
x= forest_ss.drop('size_category', axis=1)
y= forest_ss['size_category']

In [14]:
x_ss= ss.fit_transform(x)

In [15]:
x= pd.DataFrame(x_ss,columns= x.columns)
x

month       day      FFMC       DMC        DC       ISI      temp  \
0    0.271858 -1.424495 -0.796040 -1.318386 -1.828931 -0.856568 -1.827729   
1    0.958019  1.178046 -0.003257 -1.174053  0.489257 -0.507267 -0.148307   
2    0.958019 -0.383478 -0.003257 -1.043839  0.561045 -0.507267 -0.730964   
3    0.271858 -1.424495  0.194939 -1.206998 -1.896686 -0.005147 -1.810592   
4    0.271858  0.137030 -0.237488 -0.924607 -1.797070  0.125841 -1.279346   
..        ...       ...       ...       ...       ...       ...       ...   
504 -1.100464  0.137030 -1.624859 -0.839890  0.475141 -1.555170  1.531115   
505 -1.100464  0.137030 -1.624859 -0.839890  0.475141 -1.555170  0.520034   
506 -1.100464  0.137030 -1.624859 -0.839890  0.475141 -1.555170  0.400076   
507 -1.100464 -0.383478  0.681420  0.561083  0.269860  0.496973  1.154102   
508  0.729299  1.178046 -2.003232 -1.682356 -1.778921 -1.729820 -1.210798   

           RH      wind      rain  ...  monthdec  monthfeb  monthjan  \
0    0.412656  1.494370 -0.069035  ... -0.134164 -0.202237 -0.062807   
1   -0.688321 -1.729364 -0.069035  ... -0.134164 -0.202237 -0.062807   
2   -0.688321 -1.507037 -0.069035  ... -0.134164 -0.202237 -0.062807   
3    3.226265 -0.006333  0.606709  ... -0.134164 -0.202237 -0.062807   
4    3.348596 -1.229129 -0.069035  ... -0.134164 -0.202237 -0.062807   
..        ...       ...       ...  ...       ...       ...       ...   
504 -0.749487 -0.728894 -0.069035  ... -0.134164 -0.202237 -0.062807   
505  1.635965  0.994136 -0.069035  ... -0.134164 -0.202237 -0.062807   
506  1.574799  1.494370 -0.069035  ... -0.134164 -0.202237 -0.062807   
507 -0.137832 -0.006333 -0.069035  ... -0.134164 -0.202237 -0.062807   
508 -0.810652  0.271575 -0.069035  ... -0.134164 -0.202237 -0.062807   

     monthjul  monthjun  monthmar  monthmay   monthnov  monthoct  monthsep  
0    -0.25901 -0.180151  2.933219 -0.062807  -0.044368 -0.174254 -0.714413  
1    -0.25901 -0.180151 -0.340922 -0.062807  -0.044368  5.738757 -0.714413  
2    -0.25901 -0.180151 -0.340922 -0.062807  -0.044368  5.738757 -0.714413  
3    -0.25901 -0.180151  2.933219 -0.062807  -0.044368 -0.174254 -0.714413  
4    -0.25901 -0.180151  2.933219 -0.062807  -0.044368 -0.174254 -0.714413  
..        ...       ...       ...       ...        ...       ...       ...  
504  -0.25901 -0.180151 -0.340922 -0.062807  -0.044368 -0.174254 -0.714413  
505  -0.25901 -0.180151 -0.340922 -0.062807  -0.044368 -0.174254 -0.714413  
506  -0.25901 -0.180151 -0.340922 -0.062807  -0.044368 -0.174254 -0.714413  
507  -0.25901 -0.180151 -0.340922 -0.062807  -0.044368 -0.174254 -0.714413  
508  -0.25901 -0.180151 -0.340922 -0.062807  22.538855 -0.174254 -0.714413  

[509 rows x 30 columns]

In [16]:
x_train, x_test, y_train, y_test= train_test_split(x, y, test_size= 0.30, random_state= 40)

x_train.shape, x_test.shape, y_train.shape, y_test.shape

((356, 30), (153, 30), (356,), (153,))

# Grid Search CV

In [17]:
clf= SVC()
param_grid= [{'kernel':['rbf','sigmoid','poly','linear'],
               'gamma':[50,0.5,0.1,0.01],
               'C': [10,0.1,0.01,0.001,0.0001]}]
gsv= GridSearchCV(clf,param_grid,cv=10)
gsv.fit(x_train, y_train)

GridSearchCV(cv=10, estimator=SVC(),
             param_grid=[{'C': [10, 0.1, 0.01, 0.001, 0.0001],
                          'gamma': [50, 0.5, 0.1, 0.01],
                          'kernel': ['rbf', 'sigmoid', 'poly', 'linear']}])

In [18]:
gsv.best_params_, gsv.best_score_

({'C': 10, 'gamma': 50, 'kernel': 'linear'}, 0.9580158730158731)

# Support Vector Machines Classifier

In [19]:
clf= SVC(kernel='linear',C=10)
clf.fit(x_train, y_train)

y_train_pred= clf.predict(x_train)
train_acc= accuracy_score(y_train, y_train_pred)*100

print('Train Data Accuracy=', train_acc)
print(confusion_matrix(y_train, y_train_pred))


Train Data Accuracy= 98.59550561797754
[[ 98   3]
 [  2 253]]


In [20]:
#Prediction on test data.

y_test_pred=clf.predict(x_test)
test_acc= accuracy_score(y_test, y_test_pred)*100

print('Test Data Accuracy=', test_acc)
print(confusion_matrix(y_test, y_test_pred))

Test Data Accuracy= 96.73202614379085
[[ 36   1]
 [  4 112]]


In [21]:
verify= pd.DataFrame({'actual': y_test, 'predict': y_test_pred, 'equality': y_test==y_test_pred})
verify

actual  predict  equality
475       1        1      True
357       1        1      True
325       1        1      True
470       1        1      True
249       1        1      True
..      ...      ...       ...
175       1        1      True
489       1        1      True
345       1        1      True
443       1        1      True
375       0        0      True

[153 rows x 3 columns]

In [22]:
verify.equality.value_counts()

True     148
False      5
Name: equality, dtype: int64

SVM Train Accuracy = 98.6%

SVM Test Accuracy = 97%

# Problem -- 2)

Prepare a classification model using SVM for salary data

#### Data Description:

age -- age of a person

workclass -- A work class is a grouping of work

education -- Education of an individuals

maritalstatus -- Marital status of an individulas

occupation -- occupation of an individuals

relationship --

race -- Race of an Individual

sex -- Gender of an Individual

capitalgain -- profit received from the sale of an investment

capitalloss -- A decrease in the value of a capital asset

hoursperweek -- number of hours work per week

native -- Native of an individual

Salary -- salary of an individual

In [23]:
salary_train=pd.read_csv('C:/Users/ancha/Downloads/SalaryData_Train(1).csv')
salary_train

age          workclass    education  educationno        maritalstatus  \
0       39          State-gov    Bachelors           13        Never-married   
1       50   Self-emp-not-inc    Bachelors           13   Married-civ-spouse   
2       38            Private      HS-grad            9             Divorced   
3       53            Private         11th            7   Married-civ-spouse   
4       28            Private    Bachelors           13   Married-civ-spouse   
...    ...                ...          ...          ...                  ...   
30156   27            Private   Assoc-acdm           12   Married-civ-spouse   
30157   40            Private      HS-grad            9   Married-civ-spouse   
30158   58            Private      HS-grad            9              Widowed   
30159   22            Private      HS-grad            9        Never-married   
30160   52       Self-emp-inc      HS-grad            9   Married-civ-spouse   

               occupation    relationship    race      sex  capitalgain  \
0            Adm-clerical   Not-in-family   White     Male         2174   
1         Exec-managerial         Husband   White     Male            0   
2       Handlers-cleaners   Not-in-family   White     Male            0   
3       Handlers-cleaners         Husband   Black     Male            0   
4          Prof-specialty            Wife   Black   Female            0   
...                   ...             ...     ...      ...          ...   
30156        Tech-support            Wife   White   Female            0   
30157   Machine-op-inspct         Husband   White     Male            0   
30158        Adm-clerical       Unmarried   White   Female            0   
30159        Adm-clerical       Own-child   White     Male            0   
30160     Exec-managerial            Wife   White   Female        15024   

       capitalloss  hoursperweek          native  Salary  
0                0            40   United-States   <=50K  
1                0            13   United-States   <=50K  
2                0            40   United-States   <=50K  
3                0            40   United-States   <=50K  
4                0            40            Cuba   <=50K  
...            ...           ...             ...     ...  
30156            0            38   United-States   <=50K  
30157            0            40   United-States    >50K  
30158            0            40   United-States   <=50K  
30159            0            20   United-States   <=50K  
30160            0            40   United-States    >50K  

[30161 rows x 14 columns]

In [24]:
salary_test=pd.read_csv('C:/Users/ancha/Downloads/SalaryData_Test(1).csv')
salary_test

age      workclass      education  educationno        maritalstatus  \
0       25        Private           11th            7        Never-married   
1       38        Private        HS-grad            9   Married-civ-spouse   
2       28      Local-gov     Assoc-acdm           12   Married-civ-spouse   
3       44        Private   Some-college           10   Married-civ-spouse   
4       34        Private           10th            6        Never-married   
...    ...            ...            ...          ...                  ...   
15055   33        Private      Bachelors           13        Never-married   
15056   39        Private      Bachelors           13             Divorced   
15057   38        Private      Bachelors           13   Married-civ-spouse   
15058   44        Private      Bachelors           13             Divorced   
15059   35   Self-emp-inc      Bachelors           13   Married-civ-spouse   

               occupation    relationship                 race      sex  \
0       Machine-op-inspct       Own-child                Black     Male   
1         Farming-fishing         Husband                White     Male   
2         Protective-serv         Husband                White     Male   
3       Machine-op-inspct         Husband                Black     Male   
4           Other-service   Not-in-family                White     Male   
...                   ...             ...                  ...      ...   
15055      Prof-specialty       Own-child                White     Male   
15056      Prof-specialty   Not-in-family                White   Female   
15057      Prof-specialty         Husband                White     Male   
15058        Adm-clerical       Own-child   Asian-Pac-Islander     Male   
15059     Exec-managerial         Husband                White     Male   

       capitalgain  capitalloss  hoursperweek          native  Salary  
0                0            0            40   United-States   <=50K  
1                0            0            50   United-States   <=50K  
2                0            0            40   United-States    >50K  
3             7688            0            40   United-States    >50K  
4                0            0            30   United-States   <=50K  
...            ...          ...           ...             ...     ...  
15055            0            0            40   United-States   <=50K  
15056            0            0            36   United-States   <=50K  
15057            0            0            50   United-States   <=50K  
15058         5455            0            40   United-States   <=50K  
15059            0            0            60   United-States    >50K  

[15060 rows x 14 columns]

# EDA

In [25]:
salary_train.describe()

age   educationno   capitalgain   capitalloss  hoursperweek
count  30161.000000  30161.000000  30161.000000  30161.000000  30161.000000
mean      38.438115     10.121316   1092.044064     88.302311     40.931269
std       13.134830      2.550037   7406.466611    404.121321     11.980182
min       17.000000      1.000000      0.000000      0.000000      1.000000
25%       28.000000      9.000000      0.000000      0.000000     40.000000
50%       37.000000     10.000000      0.000000      0.000000     40.000000
75%       47.000000     13.000000      0.000000      0.000000     45.000000
max       90.000000     16.000000  99999.000000   4356.000000     99.000000

In [26]:
salary_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


In [27]:
salary_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15060 entries, 0 to 15059
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            15060 non-null  int64 
 1   workclass      15060 non-null  object
 2   education      15060 non-null  object
 3   educationno    15060 non-null  int64 
 4   maritalstatus  15060 non-null  object
 5   occupation     15060 non-null  object
 6   relationship   15060 non-null  object
 7   race           15060 non-null  object
 8   sex            15060 non-null  object
 9   capitalgain    15060 non-null  int64 
 10  capitalloss    15060 non-null  int64 
 11  hoursperweek   15060 non-null  int64 
 12  native         15060 non-null  object
 13  Salary         15060 non-null  object
dtypes: int64(5), object(9)
memory usage: 1.6+ MB


In [50]:
salary_train.duplicated().value_counts()

False    26903
True      3258
dtype: int64

In [55]:
#Dropping duplicates.
salary_nodup= salary_train.copy()
salary_nodup= salary_train.drop_duplicates()
salary_nodup

age          workclass      education  educationno  \
0       39          State-gov      Bachelors           13   
1       50   Self-emp-not-inc      Bachelors           13   
2       38            Private        HS-grad            9   
3       53            Private           11th            7   
4       28            Private      Bachelors           13   
...    ...                ...            ...          ...   
30154   53            Private        Masters           14   
30155   22            Private   Some-college           10   
30156   27            Private     Assoc-acdm           12   
30158   58            Private        HS-grad            9   
30160   52       Self-emp-inc        HS-grad            9   

             maritalstatus          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
30154   Married-civ-spouse     Exec-managerial         Husband   White   
30155        Never-married     Protective-serv   Not-in-family   White   
30156   Married-civ-spouse        Tech-support            Wife   White   
30158              Widowed        Adm-clerical       Unmarried   White   
30160   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capitalgain  capitalloss  hoursperweek          native  Salary  
0         Male         2174            0            40   United-States   <=50K  
1         Male            0            0            13   United-States   <=50K  
2         Male            0            0            40   United-States   <=50K  
3         Male            0            0            40   United-States   <=50K  
4       Female            0            0            40            Cuba   <=50K  
...        ...          ...          ...           ...             ...     ...  
30154     Male            0            0            40   United-States    >50K  
30155     Male            0            0            40   United-States   <=50K  
30156   Female            0            0            38   United-States   <=50K  
30158   Female            0            0            40   United-States   <=50K  
30160   Female        15024            0            40   United-States    >50K  

[26903 rows x 14 columns]

In [56]:
salary_cat=[var for var in salary_nodup.columns if salary_nodup[var].dtypes == 'O']
salary_cat= salary_nodup[salary_cat]
salary_cat.head()

workclass   education        maritalstatus          occupation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

     relationship    race      sex          native  Salary  
0   Not-in-family   White     Male   United-States   <=50K  
1         Husband   White     Male   United-States   <=50K  
2   Not-in-family   White     Male   United-States   <=50K  
3         Husband   Black     Male   United-States   <=50K  
4            Wife   Black   Female            Cuba   <=50K

In [57]:
#Frequency counts.
for var in salary_cat:
    print(salary_nodup[var].value_counts())

 Private             19213
 Self-emp-not-inc     2431
 Local-gov            2014
 State-gov            1253
 Self-emp-inc         1049
 Federal-gov           929
 Without-pay            14
Name: workclass, dtype: int64
 HS-grad         8268
 Some-college    5893
 Bachelors       4503
 Masters         1558
 Assoc-voc       1256
 Assoc-acdm       994
 11th             941
 10th             770
 7th-8th          541
 Prof-school      532
 9th              449
 Doctorate        364
 12th             358
 5th-6th          283
 1st-4th          149
 Preschool         44
Name: education, dtype: int64
 Married-civ-spouse       12371
 Never-married             8387
 Divorced                  4009
 Separated                  928
 Widowed                    817
 Married-spouse-absent      370
 Married-AF-spouse           21
Name: maritalstatus, dtype: int64
 Prof-specialty       3783
 Exec-managerial      3646
 Adm-clerical         3291
 Craft-repair         3229
 Sales                3205
 Other

In [58]:
#Find Numerical variables
salary_num=[num for num in salary_nodup.columns if salary_nodup[num].dtype !='O']
salary_nodup[salary_num].head()

age  educationno  capitalgain  capitalloss  hoursperweek
0   39           13         2174            0            40
1   50           13            0            0            13
2   38            9            0            0            40
3   53            7            0            0            40
4   28           13            0            0            40

In [59]:
#Frequency count.
for num in salary_num:
    print(salary_nodup[num].value_counts())

31    769
36    757
33    748
34    743
37    740
     ... 
82      7
83      5
88      3
85      3
86      1
Name: age, Length: 72, dtype: int64
9     8268
10    5893
13    4503
14    1558
11    1256
12     994
7      941
6      770
4      541
15     532
5      449
16     364
8      358
3      283
2      149
1       44
Name: educationno, dtype: int64
0        24379
15024      331
7688       266
7298       238
99999      147
         ...  
1639         1
6097         1
2538         1
1455         1
1086         1
Name: capitalgain, Length: 118, dtype: int64
0       25485
1902      194
1977      160
1887      153
1848       49
        ...  
2080        1
4356        1
2282        1
1844        1
1411        1
Name: capitalloss, Length: 90, dtype: int64
40    11858
50     2504
45     1642
60     1344
35     1138
      ...  
82        1
94        1
92        1
87        1
74        1
Name: hoursperweek, Length: 94, dtype: int64


# Labeling Features

In [60]:
salary_train_lab= salary_nodup.copy()

for i, column in enumerate(salary_train_lab.columns,1):
    if salary_train_lab[column].dtype!='object':
        continue
        
    label_encoder=preprocessing.LabelEncoder()
    salary_train_lab[column]=label_encoder.fit_transform(salary_train_lab[column])

In [61]:
salary_train_lab

age  workclass  education  educationno  maritalstatus  occupation  \
0       39          5          9           13              4           0   
1       50          4          9           13              2           3   
2       38          2         11            9              0           5   
3       53          2          1            7              2           5   
4       28          2          9           13              2           9   
...    ...        ...        ...          ...            ...         ...   
30154   53          2         12           14              2           3   
30155   22          2         15           10              4          10   
30156   27          2          7           12              2          12   
30158   58          2         11            9              6           0   
30160   52          3         11            9              2           3   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  \
0                 1     4    1         2174            0            40   
1                 0     4    1            0            0            13   
2                 1     4    1            0            0            40   
3                 0     2    1            0            0            40   
4                 5     2    0            0            0            40   
...             ...   ...  ...          ...          ...           ...   
30154             0     4    1            0            0            40   
30155             1     4    1            0            0            40   
30156             5     4    0            0            0            38   
30158             4     4    0            0            0            40   
30160             5     4    0        15024            0            40   

       native  Salary  
0          37       0  
1          37       0  
2          37       0  
3          37       0  
4           4       0  
...       ...     ...  
30154      37       1  
30155      37       0  
30156      37       0  
30158      37       0  
30160      37       1  

[26903 rows x 14 columns]

In [62]:
salary_test_lab= salary_test.copy()
for i, column in enumerate(salary_test_lab.columns,1):
    if salary_test_lab[column].dtype!='object':
        continue
        
    label_encoder= preprocessing.LabelEncoder()
    salary_test_lab[column]=label_encoder.fit_transform(salary_test_lab[column])

In [63]:
salary_test_lab

age  workclass  education  educationno  maritalstatus  occupation  \
0       25          2          1            7              4           6   
1       38          2         11            9              2           4   
2       28          1          7           12              2          10   
3       44          2         15           10              2           6   
4       34          2          0            6              4           7   
...    ...        ...        ...          ...            ...         ...   
15055   33          2          9           13              4           9   
15056   39          2          9           13              0           9   
15057   38          2          9           13              2           9   
15058   44          2          9           13              0           0   
15059   35          3          9           13              2           3   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  \
0                 3     2    1            0            0            40   
1                 0     4    1            0            0            50   
2                 0     4    1            0            0            40   
3                 0     2    1         7688            0            40   
4                 1     4    1            0            0            30   
...             ...   ...  ...          ...          ...           ...   
15055             3     4    1            0            0            40   
15056             1     4    0            0            0            36   
15057             0     4    1            0            0            50   
15058             3     1    1         5455            0            40   
15059             0     4    1            0            0            60   

       native  Salary  
0          37       0  
1          37       0  
2          37       1  
3          37       1  
4          37       0  
...       ...     ...  
15055      37       0  
15056      37       0  
15057      37       0  
15058      37       0  
15059      37       1  

[15060 rows x 14 columns]

# Standardization & Data Splitting

In [64]:
ss= StandardScaler()
salary_ss= salary_train_lab.copy()

In [65]:
x_train= salary_train_lab.drop('Salary',axis=1)
y_train= salary_train_lab['Salary']

x_test= salary_test_lab.drop('Salary',axis=1)
y_test= salary_test_lab['Salary']

In [66]:
x_train_ss= ss.fit_transform(x_train)
x_train= pd.DataFrame(x_train_ss,columns= x_train.columns)
x_train

age  workclass  education  educationno  maritalstatus  occupation  \
0     -0.003726   2.791383  -0.324519     1.089919       0.939836   -1.503368   
1      0.830329   1.788156  -0.324519     1.089919      -0.368576   -0.758349   
2     -0.079549  -0.218299   0.192999    -0.436737      -1.676988   -0.261669   
3      1.057799  -0.218299  -2.394592    -1.200065      -0.368576   -0.261669   
4     -0.837781  -0.218299  -0.324519     1.089919      -0.368576    0.731690   
...         ...        ...        ...          ...            ...         ...   
26898  1.057799  -0.218299   0.451758     1.471583      -0.368576   -0.758349   
26899 -1.292720  -0.218299   1.228036    -0.055073       0.939836    0.980030   
26900 -0.913604  -0.218299  -0.842037     0.708255      -0.368576    1.476709   
26901  1.436914  -0.218299   0.192999    -0.436737       2.248248   -1.503368   
26902  0.981975   0.784928   0.192999    -0.436737      -0.368576   -0.758349   

       relationship      race       sex  capitalgain  capitalloss  \
0         -0.277736  0.406417  0.705589     0.122810     -0.23133   
1         -0.895505  0.406417  0.705589    -0.155722     -0.23133   
2         -0.277736  0.406417  0.705589    -0.155722     -0.23133   
3         -0.895505 -1.889403  0.705589    -0.155722     -0.23133   
4          2.193341 -1.889403 -1.417255    -0.155722     -0.23133   
...             ...       ...       ...          ...          ...   
26898     -0.895505  0.406417  0.705589    -0.155722     -0.23133   
26899     -0.277736  0.406417  0.705589    -0.155722     -0.23133   
26900      2.193341  0.406417 -1.417255    -0.155722     -0.23133   
26901      1.575572  0.406417 -1.417255    -0.155722     -0.23133   
26902      2.193341  0.406417 -1.417255     1.769144     -0.23133   

       hoursperweek    native  
0         -0.094323  0.282414  
1         -2.281121  0.282414  
2         -0.094323  0.282414  
3         -0.094323  0.282414  
4         -0.094323 -4.979105  
...             ...       ...  
26898     -0.094323  0.282414  
26899     -0.094323  0.282414  
26900     -0.256308  0.282414  
26901     -0.094323  0.282414  
26902     -0.094323  0.282414  

[26903 rows x 13 columns]

In [67]:
x_test_ss= ss.fit_transform(x_test)
x_test= pd.DataFrame(x_test_ss, columns= x_test.columns)
x_test

age  workclass  education  educationno  maritalstatus  occupation  \
0     -1.029005  -0.222347  -2.423439    -1.216563       0.933187    0.002725   
1     -0.057423  -0.222347   0.190172    -0.434898      -0.395463   -0.494789   
2     -0.804794  -1.257126  -0.855272     0.737599      -0.395463    0.997755   
3      0.391000  -0.222347   1.235617    -0.044066      -0.395463    0.002725   
4     -0.356371  -0.222347  -2.684800    -1.607395       0.933187    0.251483   
...         ...        ...        ...          ...            ...         ...   
15055 -0.431108  -0.222347  -0.332550     1.128431       0.933187    0.748998   
15056  0.017315  -0.222347  -0.332550     1.128431      -1.724114    0.748998   
15057 -0.057423  -0.222347  -0.332550     1.128431      -0.395463    0.748998   
15058  0.391000  -0.222347  -0.332550     1.128431      -1.724114   -1.489819   
15059 -0.281634   0.812432  -0.332550     1.128431      -0.395463   -0.743547   

       relationship      race       sex  capitalgain  capitalloss  \
0          1.006108 -2.034389  0.695832    -0.145438    -0.219169   
1         -0.881990  0.382230  0.695832    -0.145438    -0.219169   
2         -0.881990  0.382230  0.695832    -0.145438    -0.219169   
3         -0.881990 -2.034389  0.695832     0.852624    -0.219169   
4         -0.252624  0.382230  0.695832    -0.145438    -0.219169   
...             ...       ...       ...          ...          ...   
15055      1.006108  0.382230  0.695832    -0.145438    -0.219169   
15056     -0.252624  0.382230 -1.437128    -0.145438    -0.219169   
15057     -0.881990  0.382230  0.695832    -0.145438    -0.219169   
15058      1.006108 -3.242699  0.695832     0.562734    -0.219169   
15059     -0.881990  0.382230  0.695832    -0.145438    -0.219169   

       hoursperweek    native  
0         -0.078889  0.260581  
1          0.750131  0.260581  
2         -0.078889  0.260581  
3         -0.078889  0.260581  
4         -0.907909  0.260581  
...             ...       ...  
15055     -0.078889  0.260581  
15056     -0.410497  0.260581  
15057      0.750131  0.260581  
15058     -0.078889  0.260581  
15059      1.579152  0.260581  

[15060 rows x 13 columns]

In [68]:
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((26903, 13), (26903,), (15060, 13), (15060,))

# Grid Search CV

In [69]:
clf= SVC()
param_grid= [{'kernel':['rbf','sigmoid','poly','linear'],
               'gamma':[0.5,0.1,0.01],
               'C': [0.1,0.01,0.001,0.0001]}]
gsv= GridSearchCV(clf,param_grid,cv=5)
gsv.fit(x_train, y_train)

GridSearchCV(cv=5, estimator=SVC(),
             param_grid=[{'C': [0.1, 0.01, 0.001, 0.0001],
                          'gamma': [0.5, 0.1, 0.01],
                          'kernel': ['rbf', 'sigmoid', 'poly', 'linear']}])

In [70]:
gsv.best_params_, gsv.best_score_

({'C': 0.1, 'gamma': 0.5, 'kernel': 'poly'}, 0.8345910469785954)

# Support Vector Machines Classifier

In [71]:
clf= SVC(kernel='poly',C=0.1)
clf.fit(x_train, y_train)

y_train_pred= clf.predict(x_train)
train_acc= accuracy_score(y_train, y_train_pred)*100

print('Train Data Accuracy=', train_acc)
print(confusion_matrix(y_train, y_train_pred))

Train Data Accuracy= 81.61915028063784
[[19542   481]
 [ 4464  2416]]


In [72]:
#Prediction on test data.

y_test_pred=clf.predict(x_test)
test_acc= accuracy_score(y_test, y_test_pred)*100

print('Test Data Accuracy=', test_acc)
print(confusion_matrix(y_test, y_test_pred))

Test Data Accuracy= 82.06507304116866
[[11075   285]
 [ 2416  1284]]


In [73]:
verify= pd.DataFrame({'actual': y_test, 'predict': y_test_pred, 'equality': y_test==y_test_pred})
verify

actual  predict  equality
0           0        0      True
1           0        0      True
2           1        0     False
3           1        0     False
4           0        0      True
...       ...      ...       ...
15055       0        0      True
15056       0        0      True
15057       0        0      True
15058       0        0      True
15059       1        0     False

[15060 rows x 3 columns]

In [74]:
verify.equality.value_counts()

True     12359
False     2701
Name: equality, dtype: int64

In [ ]:
SVM Train Accuracy= 81.6%

SVM Test Accuracy= 82%